# Harshal Talele & Harshalkumar Loya

In [198]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score

In [199]:
train_data = pd.read_csv('training_data.csv')
test_data = pd.read_csv('test_data.csv')

In [200]:
train_data['date_int'] = train_data.date_index_converted.str.replace('day_','').astype('Int64')
test_data['date_int'] = test_data.date_index_converted.str.replace('day_','').astype('Int64')

In [201]:
county_cols = ['date_int', 'county', 'cases', 'deaths', 'date_index_converted', 'county_data_length', 'total_pop', 'percent_25_34', 'percent_highschool', 'labor_force_rate', 'unemployment_rate', 'median_housing_cost', 'median_household_earnings', 'median_worker_earnings', 'percent_insured', 'percent_married', 'poverty_rate', 'median_property_value', 'percent_white']

In [202]:
corr = train_data[county_cols].corr()
corr.style.background_gradient(cmap='coolwarm').set_precision(2)

C:\Users\harsh\AppData\Local\Temp\ipykernel_3148\1680320004.py:2: FutureWarning: this method is deprecated in favour of `Styler.format(precision=..)`
  corr.style.background_gradient(cmap='coolwarm').set_precision(2)


In [203]:
len(county_cols)

19

In [204]:
county_cols_to_use = ['date_int', 'total_pop', 'deaths', 'county_data_length', 'percent_25_34', 'labor_force_rate', 'unemployment_rate', 'median_household_earnings']
len(county_cols_to_use)

8

In [205]:
social_features_string = 'core_cosine, core_cosine_normalized, core_intersection, core_intersection_nor malized, core_jaccard, core_jaccard_normalized, domestic_cosine, domestic_cos ine_normalized, domestic_intersection, domestic_intersection_normalized, dome stic_jaccard, domestic_jaccard_normalized, economy_cosine, economy_cosine_nor malized, economy_intersection, economy_intersection_normalized, economy_jacca rd, economy_jaccard_normalized, education_cosine, education_cosine_normalized , education_intersection, education_intersection_normalized, education_jaccar d, education_jaccard_normalized, entertainment_cosine, entertainment_cosine_n ormalized, entertainment_intersection, entertainment_intersection_normalized, entertainment_jaccard, entertainment_jaccard_normalized, foreign_cosine, fore ign_cosine_normalized, foreign_intersection, foreign_intersection_normalized, foreign_jaccard, foreign_jaccard_normalized, gender_cosine, gender_cosine_nor malized, gender_intersection, gender_intersection_normalized, gender_jaccard, gender_jaccard_normalized, health_cosine, health_cosine_normalized, health_in tersection, health_intersection_normalized, health_jaccard, health_jaccard_no rmalized, health_technology_cosine, health_technology_cosine_normalized, heal th_technology_intersection, health_technology_intersection_normalized, health _technology_jaccard, health_technology_jaccard_normalized, ideology_cosine, i deology_cosine_normalized, ideology_intersection, ideology_intersection_norma lized, ideology_jaccard, ideology_jaccard_normalized, illness_cosine, illness _cosine_normalized, illness_intersection, illness_intersection_normalized, il lness_jaccard, illness_jaccard_normalized, labor_force_rate, median_household _earnings, median_housing_cost, median_property_value, median_worker_earning, nationalistic_cosine, nationalistic_cosine_normalized, nationalistic_intersec tion, nationalistic_intersection_normalized, nationalistic_jaccard, nationali stic_jaccard_normalized, percent_25_34, percent_highschool, percent_insure, p ercent_married, percent_white, politics_cosine, politics_cosine_normalized, p olitics_democratic_hate_cosine, politics_democratic_hate_cosine_normalized, p olitics_democratic_hate_intersection, politics_democratic_hate_intersection_n ormalized, politics_democratic_hate_jaccard, politics_democratic_hate_jaccard _normalized, politics_democratic_love_cosine, politics_democratic_love_cosine _normalized, politics_democratic_love_intersection, politics_democratic_love_ intersection_normalized, politics_democratic_love_jaccard, politics_democrati c_love_jaccard_normalized, politics_intersection, politics_intersection_norma lized, politics_jaccard, politics_jaccard_normalized, politics_republican_hat e_cosine, politics_republican_hate_cosine_normalized, politics_republican_hat e_intersection, politics_republican_hate_intersection_normalized, politics_re publican_hate_jaccard, politics_republican_hate_jaccard_normalized, politics_ republican_love_cosine, politics_republican_love_cosine_normalized, politics_ republican_love_intersection, politics_republican_love_intersection_normalize d, politics_republican_love_jaccard, politics_republican_love_jaccard_normali zed, poverty_rate, race_cosine, race_cosine_normalized, race_intersection, ra ce_intersection_normalized, race_jaccard, race_jaccard_normalized, religion_c osine, religion_cosine_normalized, religion_intersection, religion_intersecti on_normalized, religion_jaccard, religion_jaccard_normalized, social_cosine, social_cosine_normalized, social_intersection, social_intersection_normalized , social_jaccard, social_jaccard_normalized, sports_cosine, sports_cosine_nor malized, sports_intersection, sports_intersection_normalized, sports_jaccard, sports_jaccard_normalized'

social_features_string = social_features_string.replace(" ","")
social_features = social_features_string.split(",")
social_features = [each for each in social_features_string.split(",")]

for i in range(len(social_features)):
    if social_features[i] == 'median_worker_earning':
        social_features[i] = 'median_worker_earnings'
    elif social_features[i] == 'percent_insure':
        social_features[i] = 'percent_insured'
    else:
        pass
len(social_features)

137

In [206]:
awareness_cols_to_use = ['core_intersection','health_technology_cosine_normalized', 'politics_democratic_hate_intersection', 'race_cosine']

In [207]:
cols_to_use = county_cols_to_use + awareness_cols_to_use

In [208]:
X = train_data[cols_to_use]
y = train_data['cases']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 256)

In [209]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor()
rf.fit(X_train, y_train)

y_pred_rf = rf.predict(X_test)

print('Random Forest Regression -',r2_score(y_test, y_pred_rf))

Random Forest Regression - 0.9032000839831487


In [210]:
from sklearn.ensemble import GradientBoostingRegressor

gbr = GradientBoostingRegressor()
gbr.fit(X_train, y_train)

y_pred_gbr = gbr.predict(X_test)

print('Gradient Boosting Regression -',r2_score(y_test, y_pred_gbr))

Gradient Boosting Regression - 0.9113667056127134


In [211]:
import xgboost as xg

xgb = xg.XGBRegressor()

xgb.fit(X_train, y_train)

y_pred_xgb = xgb.predict(X_test)

print('XGBoost Regression -',r2_score(y_test, y_pred_xgb))

XGBoost Regression - 0.9127768979626822


In [212]:
from sklearn.ensemble import ExtraTreesRegressor

etr = ExtraTreesRegressor()

etr.fit(X_train, y_train)

y_pred_etr = etr.predict(X_test)

print('Extra Trees Regression -',r2_score(y_test, y_pred_etr))

Extra Trees Regression - 0.9755185854707977


In [213]:
gbr.fit(X,y)
xgb.fit(X,y)
etr.fit(X,y)

ExtraTreesRegressor()

In [214]:
# gbr.fit(train_data[cols_to_use],y)
xgb.fit(train_data[cols_to_use],y)
# etr.fit(train_data[cols_to_use],y)

# test_pred = gbr.predict(test_data[cols_to_use])
test_pred = xgb.predict(test_data[cols_to_use])
# test_pred = etr.predict(test_data[cols_to_use])


out = pd.DataFrame(test_pred, columns=['Cases'])

for i in range(len(out)):
    if out['Cases'][i] < 0:
        out['Cases'][i] = 0

out['Index'] = out.index

out['Cases'] = np.floor(pd.to_numeric(out['Cases'], errors='coerce')).astype('Int64')

out = out[['Index', 'Cases']]
out

Index  Cases
0         0      1
1         1    428
2         2      0
3         3      0
4         4      0
...     ...    ...
7326   7326      0
7327   7327    103
7328   7328      0
7329   7329    517
7330   7330      0

[7331 rows x 2 columns]

In [215]:
for i in range(len(xgb.feature_names_in_)):
    print(xgb.feature_names_in_[i], "       ",xgb.feature_importances_[i]) 


date_int         0.0019541495
total_pop         0.033885073
deaths         0.5588625
county_data_length         0.028772343
percent_25_34         0.28685516
labor_force_rate         0.009185454
unemployment_rate         0.0651842
median_household_earnings         0.008705765
core_intersection         0.0009018636
health_technology_cosine_normalized         0.0010103182
politics_democratic_hate_intersection         0.004145311
race_cosine         0.00053787866


In [216]:
print('Zeros', list(out['Cases']).count(0))
print('Avg', out['Cases'].mean())
print('Std Dev', out['Cases'].std())
print(out['Cases'][34])

Zeros 4376
Avg 178.00968489974082
Std Dev 950.2225028665334
12555


In [217]:
out.to_csv('submission.csv', index=False)

In [218]:
xgb.fit(train_data[social_features],y)

XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
             interaction_constraints=None, learning_rate=None, max_bin=None,
             max_cat_threshold=None, max_cat_to_onehot=None,
             max_delta_step=None, max_depth=None, max_leaves=None,
             min_child_weight=None, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=None, num_parallel_tree=None,
             predictor=None, random_state=None, ...)

In [219]:
for i in range(len(xgb.feature_names_in_)):
    print(xgb.feature_names_in_[i], "       ",xgb.feature_importances_[i]) 


core_cosine         0.024615863
core_cosine_normalized         0.0
core_intersection         0.09556902
core_intersection_normalized         0.0
core_jaccard         0.022157174
core_jaccard_normalized         0.0
domestic_cosine         0.0021220807
domestic_cosine_normalized         0.0
domestic_intersection         0.00500702
domestic_intersection_normalized         0.0
domestic_jaccard         0.00028898945
domestic_jaccard_normalized         0.0
economy_cosine         0.00994675
economy_cosine_normalized         0.0
economy_intersection         0.0029225317
economy_intersection_normalized         0.0
economy_jaccard         0.025123214
economy_jaccard_normalized         0.0
education_cosine         0.0062960936
education_cosine_normalized         0.0
education_intersection         0.0002282981
education_intersection_normalized         0.0
education_jaccard         0.005131746
education_jaccard_normalized         0.0
entertainment_cosine         0.030499658
entertainment_cosine_nor

core_intersection         0.09556902
core_jaccard         0.022157174
labor_force_rate         0.015310176
politics_democratic_hate_intersection         0.26147783
politics_democratic_love_intersection         0.030199183
politics_republican_love_intersection         0.031491783
race_cosine         0.092853084